csrc/id_model/id_graphs.cpp

```diff
@@ -0,0 +1,1768 @@
+    // traversing the iel graph.
+    //
+    // TODO: Can we reduce the number of new expressions generated here?
+    ExprGroups non_promoted_input_uses;
+    for (auto iel_group : promoted_input_groups.intersect(input_groups)) {
+      non_promoted_input_uses.pushBack(
+          intersection_exact_loop_graph.getUniqueUses(iel_group));
+    }
+
+    for (auto iel_use_group : non_promoted_input_uses) {
+      if (IdGraph::transformAtributesMatch(
+              iel_expr->front(), iel_use_group->front())) {
+        auto use_inps =
+            ir_utils::filterByType<IterDomain>(iel_use_group->front()->inputs())
+                .vector();
+        bool inps_match = true;
+        for (auto inp_i : c10::irange(use_inps.size())) {
+          inps_match = inps_match &&
+              intersection_exact_loop_graph.disjointIdSets().strictAreMapped(
+                  use_inps[inp_i], promoted_inputs[inp_i]);
+        }
+        if (inps_match) {
+          replay = iel_use_group->front();
+          break;
+        }
+      }
+    }
+
+    bool replayed = replay == nullptr;
+    if (replay == nullptr) {
+      replay = addReplayAs(promoted_inputs, iel_expr->front());
+    }
+
+    auto out_groups = intersection_exact_loop_graph.outputGroups(iel_expr);
+
+    // Mark outputs as having a promoted iter domain
+    auto replay_out_ids =
+        ir_utils::filterByType<IterDomain>(replay->outputs()).vector();
+    auto ref_out_ids =
+        ir_utils::filterByType<IterDomain>(iel_expr->front()->outputs())
+            .vector();
+
+    TORCH_INTERNAL_ASSERT(replay_out_ids.size() == out_groups.size());
+
+    for (auto i : c10::irange(replay_out_ids.size())) {
+      iel_promotion_map[out_groups[i]] = replay_out_ids[i];
+      // Explicitly map loop map since expr propagation doesn't happen
+      if (replayed) {
+        idGraph(IdMappingMode::LOOP).mapIds(replay_out_ids[i], ref_out_ids[i]);
+      }
+    }
+  }
+  return iel_promotion_map;
+}
+
+namespace {
+
+std::unordered_map<IdGroup, IterDomain*> updateMap(
+    const std::unordered_map<IdGroup, IterDomain*> stale_map,
+    IdGraph& new_graph) {
+  std::unordered_map<IdGroup, IterDomain*> new_map;
+  for (auto stale_entry : stale_map) {
+    auto stale_id_group = stale_entry.first;
+    auto new_groups = new_graph.toGroups(*stale_id_group);
+    TORCH_INTERNAL_ASSERT(
+        new_groups.size() == 1,
+        "\nUpdate map assumes that new graph is equivalent to old graph plus extra mappings.\n",
+        "i.e. all mappings in new_graph should exist in the graph stale_map was produced on.\n",
+        "old:",
+        nvfuser::toString(stale_id_group),
+        "new: ",
+        nvfuser::toString(new_groups));
+    new_map[new_groups.front()] = stale_entry.second;
+  }
+  return new_map;
+}
+
+// Returns for each IdGroup in provided IdGraph what the input IdGroups are
+// traversing on definitions. Ignoring broadcast IdGroups and resetting inputs
+// at RFactor IdGroups.
```

Comment:
Why reset at rfactor?

Reply:
This matches what we’re doing in compute at map already today.

Reply:
I believe this is the equivalent of: https://github.com/NVIDIA/Fuser/blob/main/csrc/compute_at_map.cpp#L1046-L1071

Reply:
I don’t recall the exact sequence of logic, but we don’t have to worry about resolution of broadcast in history past an reactor shape op. I think it’s because an unresolved broadcast cannot persist through a reshape transformation because the broadcast dims can’t be merged into non-broadcast dims for something like a reshape.

Reply:
So, IIUC, no reshape should have its input and output mapped in the Permissive (and also Loop) maps since broadcast domains should be squeezed, so just terminating at rfactor ID groups should be sufficient for this analysis.

Reply:
I think that’s correct (at least for inputs/outputs that aren’t the exact same).